### PR TITLE
AP-2141 Involved Children controller

### PR DESCRIPTION
**Pull request by stephenrichards**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/AP-2141)

Add controllers and pages to allow one or more involved children to be added to the application

Added the following controllers and associated pages and forms to allow children to be added to the application using the 'add another' pattern:
- `Providers::ApplicationMeritsTask::HasOtherInvolvedChildrenController`
- `Providers::ApplicationMeritsTask::InvolvedChildrenController`
- `Providers::ApplicationMeritsTask::RemoveInvolvedChildController`

These controllers have not been included in the flow, but you can get to it by going to this URL: `/providers/applications/:legal_aid_application_id/involved_children/new`

## Checklist

Before you ask people to review this PR:

- [ ] Tests and rubocop should be passing: `bundle exec rake`
- [ ] Github should not be reporting conflicts; you should have recently run `git rebase master`.
- [ ] There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- [ ] The PR description should say what you changed and why, with a link to the JIRA story.
- [ ] You should have looked at the diff against master and ensured that nothing unexpected is included in your changes.
- [ ] You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: app/services/flow/flows/provider_merits.rb [hunk 1]
```diff
@@ -2,11 +2,40 @@ module Flow
   module Flows
     class ProviderMerits < FlowSteps
       STEPS = {
+
```
Comment (colinbruce) [inline app/services/flow/flows/provider_merits.rb:5]: New empty line

File: db/schema.rb [hunk 2]
```diff
@@ -476,7 +476,7 @@
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
     t.boolean "scanner_working"
-    t.index ["uploader_type", "uploader_id"], name: "index_malware_scan_results_on_uploader_type_and_uploader_id"
+    t.index ["uploader_type", "uploader_id"], name: "index_malware_scan_results_on_uploader"
```
Comment (colinbruce) [inline db/schema.rb:479]: Is this related to this PR?

**Review by colinbruce — COMMENTED** [review]
LGTM :+1:

A couple of minor observations but they could be picked up later when linked to the task list

File: db/schema.rb (hunk 2 above)
Comment (Alex-Swann) [reply db/schema.rb:479]: This keeps automatically happening when you run rails migrations. Mark and I have seen it time and time again. I think it's an update to clamav that's doing it (I think).

Comment (Alex-Swann) [reply db/schema.rb:479]: I think a discussion was to be had about it but has probably been forgotten. It's probably fine to leave in as it is autogenerated but might be worth a check (not sure what's the best way to do that other than everything works as expected).

File: app/forms/application_merits_task/involved_child_form.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
+module ApplicationMeritsTask
```
Comment (Alex-Swann) [inline app/forms/application_merits_task/involved_child_form.rb:1]: Might be worth moving the parent directory into the providers folder
i.e.
```
app/forms/providers/application_merits_task/involved_child_form.rb
```
Colin has done that and I was going to follow suit too.

File: app/controllers/providers/application_merits_task/involved_children_controller.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      end
+
+      def show
+        @form = ::ApplicationMeritsTask::InvolvedChildForm.new(model: involved_child)
+      end
+
+      def update
+        @form = ::ApplicationMeritsTask::InvolvedChildForm.new(form_params)
+        if @form.save
+          go_forward(involved_child)
```
Comment (Alex-Swann) [inline app/controllers/providers/application_merits_task/involved_children_controller.rb:15]:
```
return go_forward(involved_child) if @form.save
```
removes the need for the else/end after

File: app/forms/application_merits_task/involved_child_form.rb (hunk 3 above)
Comment (stephenrichards) [reply app/forms/application_merits_task/involved_child_form.rb:1]: OK, will do.

File: db/schema.rb (hunk 2 above)
Comment (stephenrichards) [reply db/schema.rb:479]: i think we can stop this happening by explicitly specifying the name - it's happening because the name is too long I think.

File: app/views/providers/application_merits_task/has_other_involved_children/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,51 @@
+<%= page_template page_title: t('.page_title', count: "#{pluralize(@legal_aid_application.involved_children.count, 'child')}"), template: :basic do %>
+
+  <% if @legal_aid_application.involved_children.count > 0 %>
+    <%= govuk_fieldset_header(size: 'l') { t('.existing', count: "#{pluralize(@legal_aid_application.involved_children.count, 'child')}")} %>
+    <div class="govuk-summary-list">
+      <table class="govuk-table">
+        <thead class="goveuk-table__head">
```
Comment (RoseSAK) [inline app/views/providers/application_merits_task/has_other_involved_children/show.html.erb:7]: random e here 'gov**e**uk-table__head' instead of 'govuk-table__head'

File: db/schema.rb (hunk 2 above)
Comment (Alex-Swann) [reply db/schema.rb:479]: I think this has been updated elsewhere now doing precisely that. So if you rebase I think this will then disappear from the PR anyway.

File: app/views/providers/application_merits_task/has_other_involved_children/show.html.erb (hunk 5 above)
Comment (stephenrichards) [reply app/views/providers/application_merits_task/has_other_involved_children/show.html.erb:7]: Eagle-eyed!  👍

File: spec/factories/application_merits_task/involved_children.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+module ApplicationMeritsTask
```
Comment (Alex-Swann) [inline spec/factories/application_merits_task/involved_children.rb:1]: Sorry, boring comment. Does this need to be pluralised along with the parent dir for consistency? I know we do this with the CFE Results factory so just checking.

File: spec/forms/providers/application_merits_task/involved_child_form_spec.rb [hunk 7]
```diff
@@ -0,0 +1,68 @@
+require 'rails_helper'
+
+module Providers
+  module ApplicationMeritsTask
```
Comment (Alex-Swann) [inline spec/forms/providers/application_merits_task/involved_child_form_spec.rb:4]: same question here

File: app/forms/providers/application_merits_task/involved_child_form.rb [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+      attr_accessor(*ATTRIBUTES)
+      attr_writer :date_of_birth
+
+      validates :full_name, presence: true
+      validates :date_of_birth, presence: true
+      validates(
+        :date_of_birth,
+        date: {
+          not_in_the_future: true,
+          earliest_allowed_date: { date: '2000-01-01' }
```
Comment (willc-work) [inline app/forms/providers/application_merits_task/involved_child_form.rb:24]: Is this the actual earliest allowed date? This allows for someone who is 21 years old. We may need to make this date dynamic based on the current date and what we consider a child.

File: spec/forms/providers/application_merits_task/involved_child_form_spec.rb [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            expect(subject).not_to be_valid
+            expect(subject.errors[:full_name]).to eq ["can't be blank"]
+          end
+        end
+
+        context 'date too early' do
+          let(:dob) { earliest_date - 2.years }
+          it 'returns false' do
+            expect(subject).not_to be_valid
+            expect(subject.errors[:date_of_birth]).to eq ['Date cannot be before 01 01 2000']
```
Comment (willc-work) [inline spec/forms/providers/application_merits_task/involved_child_form_spec.rb:39]: If the age limit is changed in validation this will need to change

**Review by willc-work — APPROVED** [review]
